main.c

```diff
@@ -235,18 +236,78 @@ static bool path_exists(const char *path) {
 	return path && access(path, R_OK) != -1;
 }
 
-static const char *get_output_dir(void) {
-	static const char *output_dirs[] = {
-		"GRIM_DEFAULT_DIR",
-		"XDG_PICTURES_DIR",
-	};
+char *get_xdg_pictures_dir(void) {
+	char *home_dir = getenv("HOME");
+	if (home_dir == NULL) {
+		return NULL;
+	}
+
+	char *config_file;
+	char *config_home = getenv("XDG_CONFIG_HOME");
+	if (config_home == NULL || config_home[0] == 0) {
+		config_file = (char *)malloc(strlen(home_dir) + strlen("/.config/user-dirs.dirs") + 1);
+		if (config_file == NULL) {
+			return NULL;
+		}
+		strcpy(config_file, home_dir);
+		strcat(config_file, "/.config/user-dirs.dirs");
+	} else {
+		config_file = (char *)malloc(strlen(config_home) + strlen("/user-dirs.dirs") + 1);
+		if (config_file == NULL) {
+			return NULL;
+		}
+		strcpy(config_file, config_home);
+		strcat(config_file, "/user-dirs.dirs");
+	}
+
+	FILE *file = fopen(config_file, "r");
+	free(config_file);
+	if (file == NULL) {
+		return NULL;
+	}
+
+	char *line = NULL;
+	size_t line_size = 0;
+	ssize_t nread;
+	char *pictures_dir = NULL;
+	while ((nread = getline(&line, &line_size, file)) != -1) {
+		if (line[nread - 1] == '\n') {
+			line[--nread] = '\0';
+		}
 
-	for (size_t i = 0; i < sizeof(output_dirs) / sizeof(char *); ++i) {
-		const char *path = getenv(output_dirs[i]);
-		if (path_exists(path)) {
-			return path;
+		if (strlen(line) == 0 || line[0] == '#') {
+			continue;
+		}
+
+		size_t i = 0;
+		while (line[i] != '\0' && line[i] != ' ') {
+			i++;
+		}
+		if (strncmp(line, "XDG_PICTURES_DIR=", 17) == 0) {
+			char *line_remaining = line + 17;
+			wordexp_t p;
+			if (wordexp(line_remaining, &p, WRDE_UNDEF) == 0) {
+				free(pictures_dir);
+				pictures_dir = strdup(p.we_wordv[0]);
+				wordfree(&p);
+			}
 		}
 	}
+	free(line);
+	fclose(file);
+	return pictures_dir;
+}
+
+char *get_output_dir(void) {
+	char *grim_default_dir = getenv("GRIM_DEFAULT_DIR");
+	if (path_exists(grim_default_dir)) {
+		return grim_default_dir;
+	}
+
+	char *xdg_fallback_dir = get_xdg_pictures_dir();
+	if (path_exists(xdg_fallback_dir)) {
+		return xdg_fallback_dir;
+	}
 
 	return ".";
 }
```

Comment:
> Thanks, I just tested with ASan as well and there should be no more leaks now.

Not along the standard path when `get_xdg_pictures_dir()` returns something, but:
* If `grim_default_dir` is nonzero, it will later be freed, even though it wasn't heap allocated. `strdup(grim_default_dir)` might fix this. The same need to `strdup` holds for `return ".";`
* If `path_exists` fails, then `xdg_fallback_dir` will still be leaked.

Reply:
Thanks again! I ran through those code-paths with the latest commits and there were no complains from ASan.